### PreProcessing Reads

In [1]:
!amptk ion -i mock3.ion.fastq --barcode_fasta mock3.barcodes.fa --min_len 100 -o ion3 -l 300

-------------------------------------------------------
[08:48:15 PM]: OS: MacOSX 10.12.5, 8 cores, ~ 16 GB RAM. Python: 2.7.12
[08:48:15 PM]: AMPtk v.0.10.0, USEARCH v9.2.64, VSEARCH v2.4.3
[08:48:16 PM]: Foward primer: AGTGARTCATCGAATCTTTG,  Rev comp'd rev primer: GCATATCAATAAGCGGAGGA
[08:48:16 PM]: Loading FASTQ Records
[08:48:21 PM]: 4,332,502 reads (3.3 GB)
-------------------------------------------------------
[08:51:11 PM]: Concatenating Demuxed Files
[08:51:46 PM]: 4,332,502 total reads
[08:51:46 PM]: 3,115,600 valid Barcode
[08:51:46 PM]: 3,091,409 Fwd Primer found, 2,544,919 Rev Primer found
[08:51:46 PM]: 67,482 discarded too short (< 100 bp)
[08:51:46 PM]: 3,023,927 valid output reads
[08:51:56 PM]: Found 19 barcoded samples
                Sample:  Count
              m3-766-1:  216946
       m3-mock3-32000b:  195110
        m3-mock3-16000:  190562
        m3-mock3-3200a:  188878
              m3-301-2:  177773
              m3-744-2:  167910
              m3-500-1:  1649

In [2]:
#Now process first illumina run containing termX samples + biomock
!amptk illumina -i termX_miseqA -o m4a --min_len 100 --read_length 300 --cleanup -l 300

-------------------------------------------------------
[08:51:57 PM]: OS: MacOSX 10.12.5, 8 cores, ~ 16 GB RAM. Python: 2.7.12
[08:51:57 PM]: AMPtk v.0.10.0, USEARCH v9.2.64, VSEARCH v2.4.3
[08:51:57 PM]: Merging Overlaping Pairs using USEARCH
[08:51:57 PM]: working on sample m4A-301-1 (Read Length: 300)
[08:52:10 PM]: 270,478 reads passed (100.0%)
[08:52:10 PM]: working on sample m4A-301-2 (Read Length: 300)
[08:52:22 PM]: 296,225 reads passed (100.0%)
[08:52:22 PM]: working on sample m4A-500-1 (Read Length: 300)
[08:52:37 PM]: 334,852 reads passed (100.0%)
[08:52:37 PM]: working on sample m4A-500-2 (Read Length: 300)
[08:52:50 PM]: 275,990 reads passed (100.0%)
[08:52:50 PM]: working on sample m4A-712-1 (Read Length: 300)
[08:53:01 PM]: 265,983 reads passed (100.0%)
[08:53:01 PM]: working on sample m4A-712-2 (Read Length: 300)
[08:53:13 PM]: 242,159 reads passed (100.0%)
[08:53:13 PM]: working on sample m4A-736-1 (Read Length: 300)
[08:53:26 PM]: 279,894 reads passed (100.0%)
[08:53

In [3]:
#Now process first illumina run containing termX samples + biomock but run with 1 mismatch binning
!amptk illumina -i miseqA_1mm -o m4a1m --min_len 100 --read_length 300 --cleanup -l 300

-------------------------------------------------------
[08:57:45 PM]: OS: MacOSX 10.12.5, 8 cores, ~ 16 GB RAM. Python: 2.7.12
[08:57:45 PM]: AMPtk v.0.10.0, USEARCH v9.2.64, VSEARCH v2.4.3
[08:57:45 PM]: Merging Overlaping Pairs using USEARCH
[08:57:45 PM]: working on sample m4a-1mm-301-1 (Read Length: 300)
[08:57:58 PM]: 283,491 reads passed (100.0%)
[08:57:58 PM]: working on sample m4a-1mm-301-2 (Read Length: 300)
[08:58:12 PM]: 310,134 reads passed (100.0%)
[08:58:12 PM]: working on sample m4a-1mm-500-1 (Read Length: 300)
[08:58:28 PM]: 351,168 reads passed (100.0%)
[08:58:28 PM]: working on sample m4a-1mm-500-2 (Read Length: 300)
[08:58:41 PM]: 289,626 reads passed (100.0%)
[08:58:41 PM]: working on sample m4a-1mm-712-1 (Read Length: 300)
[08:58:53 PM]: 277,126 reads passed (100.0%)
[08:58:53 PM]: working on sample m4a-1mm-712-2 (Read Length: 300)
[08:59:05 PM]: 259,413 reads passed (100.0%)
[08:59:05 PM]: working on sample m4a-1mm-736-1 (Read Length: 300)
[08:59:17 PM]: 293,621 

In [4]:
#process illumina run B; this was the same 4 libraries run on a different run, shows no index bleed
!amptk illumina -i termX_miseqB -o m4b --min_len 100 --read_length 300 --cleanup -l 300

-------------------------------------------------------
[09:03:50 PM]: OS: MacOSX 10.12.5, 8 cores, ~ 16 GB RAM. Python: 2.7.12
[09:03:51 PM]: AMPtk v.0.10.0, USEARCH v9.2.64, VSEARCH v2.4.3
[09:03:51 PM]: Merging Overlaping Pairs using USEARCH
[09:03:51 PM]: working on sample m4b-744-1 (Read Length: 300)
[09:03:59 PM]: 191,554 reads passed (100.0%)
[09:03:59 PM]: working on sample m4b-744-2 (Read Length: 300)
[09:04:06 PM]: 156,694 reads passed (100.0%)
[09:04:06 PM]: working on sample m4b-766-1 (Read Length: 300)
[09:04:14 PM]: 165,809 reads passed (100.0%)
[09:04:14 PM]: working on sample m4b-766-2 (Read Length: 300)
[09:04:21 PM]: 145,681 reads passed (100.0%)
[09:04:21 PM]: Stripping primers and trim to 300 bp
[09:04:21 PM]: splitting the job over 8 cpus, but this may still take awhile
[09:04:21 PM]: Foward primer: GTGARTCATCGAATCTTTG,  Rev comp'd rev primer: GCATATCAATAAGCGGAGGA
-------------------------------------------------------
[09:04:29 PM]: Concatenating Demuxed Files
[09

In [5]:
#now demux miseq run 3, remade libraries from same amplicons with different barcodes
!amptk illumina -i termX_miseqC -o m4c --min_len 100 --read_length 300 --cleanup -l 300

-------------------------------------------------------
[09:04:33 PM]: OS: MacOSX 10.12.5, 8 cores, ~ 16 GB RAM. Python: 2.7.12
[09:04:34 PM]: AMPtk v.0.10.0, USEARCH v9.2.64, VSEARCH v2.4.3
[09:04:34 PM]: Merging Overlaping Pairs using USEARCH
[09:04:34 PM]: working on sample m4C-301-1 (Read Length: 300)
[09:04:48 PM]: 314,811 reads passed (100.0%)
[09:04:48 PM]: working on sample m4C-301-2 (Read Length: 300)
[09:05:05 PM]: 388,350 reads passed (100.0%)
[09:05:05 PM]: working on sample m4C-500-1 (Read Length: 300)
[09:05:18 PM]: 300,487 reads passed (100.0%)
[09:05:18 PM]: working on sample m4C-500-2 (Read Length: 300)
[09:05:31 PM]: 298,857 reads passed (100.0%)
[09:05:31 PM]: working on sample m4C-712-1 (Read Length: 300)
[09:05:45 PM]: 307,691 reads passed (100.0%)
[09:05:45 PM]: working on sample m4C-712-2 (Read Length: 300)
[09:05:59 PM]: 302,846 reads passed (100.0%)
[09:05:59 PM]: working on sample m4C-736-1 (Read Length: 300)
[09:06:13 PM]: 290,077 reads passed (100.0%)
[09:06

In [6]:
#now finally one more run, where we used SynMock on same samples using Ion torrent
!amptk ion -i mock5.ion.fastq --barcode_fasta mock5.barcodes.fa --min_len 100 -o ion5 -l 300

-------------------------------------------------------
[09:10:48 PM]: OS: MacOSX 10.12.5, 8 cores, ~ 16 GB RAM. Python: 2.7.12
[09:10:48 PM]: AMPtk v.0.10.0, USEARCH v9.2.64, VSEARCH v2.4.3
[09:10:48 PM]: Foward primer: AGTGARTCATCGAATCTTTG,  Rev comp'd rev primer: GCATATCAATAAGCGGAGGA
[09:10:48 PM]: Loading FASTQ Records
[09:10:53 PM]: 4,341,392 reads (3.3 GB)
-------------------------------------------------------
[09:14:05 PM]: Concatenating Demuxed Files
[09:14:44 PM]: 4,341,392 total reads
[09:14:44 PM]: 3,230,560 valid Barcode
[09:14:44 PM]: 3,204,111 Fwd Primer found, 2,790,457 Rev Primer found
[09:14:44 PM]: 52,420 discarded too short (< 100 bp)
[09:14:44 PM]: 3,151,691 valid output reads
[09:14:54 PM]: Found 24 barcoded samples
                Sample:  Count
   m5-SynMock1-mock3.0:  294636
              m5-712-2:  254027
              m5-301-2:  201097
m5-SynMock1.736-1.736-2:  187364
              m5-301-1:  187165
              m5-500-1:  172804
      m5-SynMock1-8000:  159

In [76]:
#demux the 4th miseq run - which has SynMock added, will remove barcode mismatches
!amptk illumina -i mock6 -o m6 --min_len 100 --read_length 300 --cleanup -l 300 --barcode_mismatch 0

-------------------------------------------------------
[05:01:08 PM]: OS: MacOSX 10.12.5, 8 cores, ~ 16 GB RAM. Python: 2.7.12
[05:01:08 PM]: AMPtk v.0.10.0, USEARCH v9.2.64, VSEARCH v2.4.3
[05:01:08 PM]: Merging Overlaping Pairs using USEARCH
[05:01:08 PM]: working on sample m6-301-1 (Read Length: 300)
^C
Traceback (most recent call last):
  File "/Users/jon/amptk/amptk", line 2, in <module>
    import amptk
  File "/Users/jon/amptk/amptk.py", line 220, in <module>
    subprocess.call(arguments)
  File "/Users/jon/miniconda2/lib/python2.7/subprocess.py", line 523, in call
    return Popen(*popenargs, **kwargs).wait()
  File "/Users/jon/miniconda2/lib/python2.7/subprocess.py", line 1392, in wait
Traceback (most recent call last):
  File "/Users/jon/amptk/bin/amptk-process_illumina_folder.py", line 285, in <module>
    amptklib.MergeReads(for_reads, rev_reads, args.out, name+'.fq', read_length, args.min_len, args.usearch, args.rescue_forward, args.merge_method, Index, args.barcode_mism

In [77]:
#demux the 4th miseq run but leave 1 mismatch in index reads
!amptk illumina -i mock6 -o m6-1mm --min_len 100 --read_length 300 --cleanup -l 300

-------------------------------------------------------
[05:01:52 PM]: OS: MacOSX 10.12.5, 8 cores, ~ 16 GB RAM. Python: 2.7.12
[05:01:53 PM]: AMPtk v.0.10.0, USEARCH v9.2.64, VSEARCH v2.4.3
[05:01:53 PM]: Merging Overlaping Pairs using USEARCH
[05:01:53 PM]: working on sample m6-301-1 (Read Length: 300)
[05:02:35 PM]: 890,433 reads passed (100.0%)
[05:02:35 PM]: working on sample m6-301-2 (Read Length: 300)
[05:03:31 PM]: 1,181,538 reads passed (100.0%)
[05:03:31 PM]: working on sample m6-500-1 (Read Length: 300)
[05:04:16 PM]: 911,764 reads passed (100.0%)
[05:04:16 PM]: working on sample m6-500-2 (Read Length: 300)
[05:04:59 PM]: 879,717 reads passed (100.0%)
[05:04:59 PM]: working on sample m6-712-1 (Read Length: 300)
[05:05:42 PM]: 897,131 reads passed (100.0%)
[05:05:42 PM]: working on sample m6-712-2 (Read Length: 300)
[05:06:25 PM]: 872,239 reads passed (100.0%)
[05:06:25 PM]: working on sample m6-736-1 (Read Length: 300)
[05:07:12 PM]: 998,788 reads passed (100.0%)
[05:07:12 P

In [30]:
#concatenate all miseq runs
!cat m4a.demux.fq m4b.demux.fq m4c.demux.fq m6.demux.fq > miseq.fq

In [31]:
!amptk show -i miseq.fq

----------------------------------
Found 65 barcoded samples
              Sample:  Count
          m6-SynMock:  1167383
            m6-757-2:  1099049
            m6-301-2:  1059574
     m6-mock3-32000a:  990581
            m6-755-2:  988744
            m6-736-1:  902612
            m6-755-1:  870731
     m6-mock3-32000b:  843586
            m6-736-2:  839082
            m6-500-1:  831847
            m6-757-1:  825361
      m6-mock3-16000:  818345
            m6-301-1:  815157
            m6-712-1:  812102
            m6-500-2:  808875
            m6-712-2:  795405
            m6-766-1:  759364
             m6-stds:  713488
            m6-766-2:  704442
            m6-744-1:  703396
            m6-744-2:  622853
           m4C-757-2:  392019
           m4C-301-2:  388026
           m4C-755-2:  353925
    m4C-mock3-32000a:  346977
           m4A-500-1:  333978
           m4A-766-2:  323283
           m4C-736-2:  316678
           m4C-301-1:  314526
           m4C-757-1:  312058
       

#### Run UPARSE on entire Miseq dataset and then Ion datasets

In [32]:
#cluster the miseq data
!amptk cluster -i miseq.fq -o miseq

-------------------------------------------------------
[10:51:49 AM]: OS: MacOSX 10.12.5, 8 cores, ~ 16 GB RAM. Python: 2.7.12
[10:51:50 AM]: AMPtk v.0.10.0, USEARCH v9.2.64, VSEARCH v2.4.3
[10:51:50 AM]: Loading FASTQ Records
[10:52:18 AM]: 30,378,249 reads (16.7 GB)
[10:52:18 AM]: Quality Filtering, expected errors < 1.0
[11:16:20 AM]: 24,508,786 reads passed
[11:16:20 AM]: De-replication (remove duplicate reads)
[11:18:25 AM]: 819,429 reads passed
[11:18:28 AM]: Clustering OTUs (UPARSE)
[11:19:45 AM]: 2,816 OTUs
[11:19:45 AM]: Cleaning up padding from OTUs
[11:19:46 AM]: Mapping Reads to OTUs and Building OTU table
[12:47:28 PM]: 26,392,942 reads mapped to OTUs (87%)
-------------------------------------------------------
OTU Clustering Script has Finished Successfully
-------------------------------------------------------
Clustered OTUs: /Users/jon/projects/synmock_paper/miseq.cluster.otus.fa
OTU Table: /Users/jon/projects/synmock_paper/miseq.otu_table.txt
-----------------------

In [33]:
#get taxonomy for miseq OTUs
!amptk taxonomy -f miseq.cluster.otus.fa -o miseq -d ITS2

-------------------------------------------------------
[12:50:20 PM]: OS: MacOSX 10.12.5, 8 cores, ~ 16 GB RAM. Python: 2.7.12
[12:50:20 PM]: AMPtk v.0.10.0, USEARCH v9.2.64, VSEARCH v2.4.3
[12:50:20 PM]: Loading FASTA Records
[12:50:20 PM]: 2,816 OTUs
[12:50:20 PM]: Global alignment OTUs with usearch_global (USEARCH)
[12:51:04 PM]: Classifying OTUs with UTAX (USEARCH)
[12:51:06 PM]: Classifying OTUs with SINTAX (USEARCH)
[12:59:10 PM]: Generating phylogenetic tree
[12:59:34 PM]: Taxonomy finished: miseq.taxonomy.txt
[12:59:34 PM]: OTUs with taxonomy: miseq.otus.taxonomy.fa
[12:59:34 PM]: OTU phylogeny: miseq.tree.phy
-------------------------------------------------------


In [12]:
#concatenate all ion runs
!cat ion3.demux.fq ion5.demux.fq > ion.fq

In [13]:
#cluster the ion data
!amptk cluster -i ion.fq -o ion

-------------------------------------------------------
[12:15:01 AM]: OS: MacOSX 10.12.5, 8 cores, ~ 16 GB RAM. Python: 2.7.12
[12:15:02 AM]: AMPtk v.0.10.0, USEARCH v9.2.64, VSEARCH v2.4.3
[12:15:02 AM]: Loading FASTQ Records
[12:15:08 AM]: 6,175,618 reads (3.4 GB)
[12:15:08 AM]: Quality Filtering, expected errors < 1.0
[12:17:39 AM]: 2,688,022 reads passed
[12:17:39 AM]: De-replication (remove duplicate reads)
[12:17:49 AM]: 109,495 reads passed
[12:17:49 AM]: Clustering OTUs (UPARSE)
[12:18:00 AM]: 1,310 OTUs
[12:18:00 AM]: Cleaning up padding from OTUs
[12:18:00 AM]: Mapping Reads to OTUs and Building OTU table
[12:34:03 AM]: 4,943,835 reads mapped to OTUs (80%)
-------------------------------------------------------
OTU Clustering Script has Finished Successfully
-------------------------------------------------------
Clustered OTUs: /Users/jon/projects/synmock_paper/ion.cluster.otus.fa
OTU Table: /Users/jon/projects/synmock_paper/ion.otu_table.txt
-------------------------------

In [14]:
#get taxonomy for ion OTUs
!amptk taxonomy -f ion.cluster.otus.fa -o ion -d ITS2

-------------------------------------------------------
[12:34:04 AM]: OS: MacOSX 10.12.5, 8 cores, ~ 16 GB RAM. Python: 2.7.12
[12:34:04 AM]: AMPtk v.0.10.0, USEARCH v9.2.64, VSEARCH v2.4.3
[12:34:04 AM]: Loading FASTA Records
[12:34:04 AM]: 1,310 OTUs
[12:34:04 AM]: Global alignment OTUs with usearch_global (USEARCH)
[12:34:21 AM]: Classifying OTUs with UTAX (USEARCH)
[12:34:22 AM]: Classifying OTUs with SINTAX (USEARCH)
[12:37:46 AM]: Generating phylogenetic tree
[12:37:52 AM]: Taxonomy finished: ion.taxonomy.txt
[12:37:52 AM]: OTUs with taxonomy: ion.otus.taxonomy.fa
[12:37:52 AM]: OTU phylogeny: ion.tree.phy
-------------------------------------------------------


#### Try to Measure index bleed based on the mock communities

In [46]:
#measure index bleed
#clean up the miseq and ion combined runs to look for index bleed, filter without removing anything
!amptk filter -i miseq.otu_table.txt -f miseq.cluster.otus.fa -p 0 -o miseq -b m4A-stds -m mock2 --keep_mock
!amptk filter -i ion.otu_table.txt -f ion.cluster.otus.fa -p 0 -o ion -b m3-stds -m mock2 --keep_mock
!amptk filter -i miseq.otu_table.txt -f miseq.cluster.otus.fa -p 0 -o miseq_syn -b m6-SynMock -m synmock --keep_mock
!amptk filter -i ion.otu_table.txt -f ion.cluster.otus.fa -p 0 -o ion_syn -b m5-SynMock2-8000 -m synmock --keep_mock

-------------------------------------------------------
[01:36:23 PM]: OS: MacOSX 10.12.5, 8 cores, ~ 16 GB RAM. Python: 2.7.12
[01:36:23 PM]: AMPtk v.0.10.0, USEARCH v9.2.64, VSEARCH v2.4.3
[01:36:23 PM]: Loading OTU table: miseq.otu_table.txt
[01:36:23 PM]: OTU table contains 2816 OTUs
[01:36:23 PM]: Mapping OTUs to Mock Community (USEARCH)
[01:36:23 PM]: Sorting OTU table naturally
[01:36:23 PM]: Removing OTUs according to --min_reads_otu: (OTUs with less than 2 reads from all samples)
[01:36:23 PM]: Normalizing OTU table to number of reads per sample
[01:36:23 PM]: Index bleed, mock into samples: 91.850853%.  Index bleed, samples into mock: 0.360022%.
[01:36:23 PM]: Overwriting auto detect index-bleed, setting to 0.000000%
[01:36:28 PM]: Filtering OTU table down to 2471 OTUs
[01:36:28 PM]: Filtering valid OTUs
-------------------------------------------------------
OTU Table filtering finished
-------------------------------------------------------
OTU Table Stats:      miseq.stats

In [47]:
#need to measure index bleed, lets start with ion dataset
import pandas as pd
import numpy as np
from natsort import natsorted
#now load sorted "raw" OTU table into dataframe
df = pd.read_csv('ion.final.txt', sep='\t', index_col=0)
#print list(df.columns.values)
#print len(df)
#okay, so 1310 OTUs in entire data set, lets trim this down to only run 3, as that one is labeled here with biomock
keepers = ['m3-301-1', 'm3-301-2', 'm3-500-1', 'm3-500-2', 'm3-712-1', 'm3-712-2', 
           'm3-736-1', 'm3-736-2', 'm3-744-1', 'm3-744-2', 'm3-755-1', 
           'm3-757-2', 'm3-766-1', 'm3-766-2', 'm3-mock3-3200a', 'm3-mock3-16000', 'm3-mock3-32000b', 'm3-stds']
mock_samples = ['m3-mock3-3200a', 'm3-mock3-16000', 'm3-mock3-32000b', 'm3-stds']
samples = [item for item in keepers if item not in mock_samples]
#print samples
df2 = df.reindex(columns=keepers, index=natsorted(df.index))
df2.to_csv('ion3.bleed.otu_table.txt', sep='\t')

#now drop OTUs where they are all zero
df3 = df2.loc[(df2 != 0).any(1)]
#print df3
mock_otus = []
for x in df3.index:
    if x.startswith('mock'):
        if not x.startswith('mock3'): #mock3 is actually in some of the samples so can't use for index-bleed measurment
            mock_otus.append(x)
#print mock_otus
#now slice into two tables containing only mocks and calculate index-bleed
bleed_counts = df3.reindex(columns=samples, index=mock_otus)
bleed_sum = bleed_counts.values.sum()
real_counts = df3.reindex(columns=mock_samples, index=mock_otus)
real_sum = real_counts.values.sum()
index_bleed = int(bleed_sum) / float(real_sum) * 100
print "Index-bleed out of mock: %f%% (%i out of %i)" % (index_bleed, bleed_sum, real_sum)

#now try to to do the opposite, count number of times that samples move into mock
#first generate list of mock otus
sample_otus = []
for x in df3.index:
    if not x.startswith('mock'):
        sample_otus.append(x)
#slice first to samples, further trim list of only things found in mocks, would be spurious or chimeras
sampledf = df3.reindex(columns=samples, index=sample_otus)
sampledf = sampledf.loc[(sampledf != 0).any(1)]
cleaned_otus = sampledf.index
#print len(cleaned_otus), len(sample_otus)

#now using new list grab counts in mock
bleed_counts = df3.reindex(columns=mock_samples, index=cleaned_otus)
bleed_counts = bleed_counts.loc[(bleed_counts != 0).any(1)]
#print bleed_counts
bleed_sum = bleed_counts.values.sum()
#print len(bleed_counts.index), bleed_sum

#finally go back to samples and get read counts for the ovlerlapping otus
real_counts = df3.reindex(columns=samples, index=bleed_counts.index)
real_sum = real_counts.values.sum()
index_bleed2 = int(bleed_sum) / float(real_sum) * 100
print "Index-bleed into mock: %f%% (%i out of %i)" % (index_bleed2, bleed_sum, real_sum)
print "Total index-bleed: %f%%" % (index_bleed + index_bleed2)

Index-bleed out of mock: 0.025337% (114 out of 449930)
Index-bleed into mock: 0.018041% (191 out of 1058702)
Total index-bleed: 0.043378%


In [48]:
#now try on the ion5 dataset
#need to measure index bleed, lets start with ion dataset
import pandas as pd
import numpy as np
from natsort import natsorted
import sys
#now load sorted "raw" OTU table into dataframe
df = pd.read_csv('ion_syn.final.txt', sep='\t', index_col=0)
#print list(df.columns.values)
#sys.exit(1)
#print len(df)
#okay, so 1367 OTUs in entire data set, lets trim this down to only run 3, as that one is labeled here with biomock
keepers = ['m5-301-1', 'm5-301-2', 'm5-500-1', 'm5-500-2', 'm5-712-1', 'm5-712-2', 'm5-736-1', 'm5-736-2', 
           'm5-744-1', 'm5-744-2', 'm5-755-1', 'm5-755-2', 'm5-757-1', 'm5-757-2', 'm5-766-1', 'm5-766-2', 
           'm5-SynMock1-8000', 'm5-SynMock1-mock3.0', 'm5-SynMock1-mock3.5', 'm5-SynMock1.736-1.736-2', 
           'm5-SynMock2-8000', 'm5-SynMock2.757-1.757-2', 'm5-mock3.0-16000', 'm5-mock3.5-16000']
mock_samples = ['m5-SynMock2-8000', 'm5-SynMock2.757-1.757-2', 'm5-SynMock1-8000', 
                'm5-SynMock1-mock3.0', 'm5-SynMock1-mock3.5', 'm5-SynMock1.736-1.736-2']
samples = [item for item in keepers if item not in mock_samples]
#print samples
df2 = df.reindex(columns=keepers, index=natsorted(df.index))

#now drop OTUs where they are all zero
df3 = df2.loc[(df2 != 0).any(1)]
mock_otus = []
for x in df3.index:
    if 'mock' in x: #note this is different here as I'm also including mock chimeras as based on synthetic mock
        mock_otus.append(x)
#now slice into two tables containing only mocks and calculate index-bleed
bleed_counts = df3.reindex(columns=samples, index=mock_otus)
bleed_sum = bleed_counts.values.sum()
real_counts = df3.reindex(columns=mock_samples, index=mock_otus)
real_sum = real_counts.values.sum()
index_bleed = int(bleed_sum) / float(real_sum) * 100
print "Index-bleed out of mock: %f%% (%i out of %i)" % (index_bleed, bleed_sum, real_sum)

#now try to to do the opposite, count number of times that samples move into mock
#first generate list of mock otus
sample_otus = []
for x in df3.index:
    if not 'mock' in x: #note this is different here as I'm also including mock chimeras as based on synthetic mock
        sample_otus.append(x)
clean_mocks = ['m5-SynMock2-8000', 'm5-SynMock1-8000',]
#slice first to samples, further trim list of only things found in mocks, would be spurious or chimeras
sampledf = df3.reindex(columns=samples, index=sample_otus)
sampledf = sampledf.loc[(sampledf != 0).any(1)]
cleaned_otus = sampledf.index
#print len(cleaned_otus), len(sample_otus)

#now using new list grab counts in mock
bleed_counts = df3.reindex(columns=clean_mocks, index=cleaned_otus)
bleed_counts = bleed_counts.loc[(bleed_counts != 0).any(1)]
bleed_sum = bleed_counts.values.sum()
#print len(bleed_counts.index), bleed_sum

#finally go back to samples and get read counts for the ovlerlapping otus
real_counts = df3.reindex(columns=samples, index=bleed_counts.index)
real_sum = real_counts.values.sum()
index_bleed2 = int(bleed_sum) / float(real_sum) * 100
print "Index-bleed into mock: %f%% (%i out of %i)" % (index_bleed2, bleed_sum, real_sum)
print "Total index-bleed: %f%%" % (index_bleed + index_bleed2)

Index-bleed out of mock: 0.040805% (226 out of 553853)
Index-bleed into mock: 0.028418% (91 out of 320223)
Total index-bleed: 0.069223%


In [49]:
#need to measure index bleed of miseq data now, trying 0 mismatch miseqA dataset
import pandas as pd
import numpy as np
from natsort import natsorted
import sys
#now load sorted "raw" OTU table into dataframe
df = pd.read_csv('miseq.final.txt', sep='\t', index_col=0)
#print list(df.columns.values)
#sys.exit(1)
#print len(df)
#okay, so 1367 OTUs in entire data set, lets trim this down to only run 3, as that one is labeled here with biomock
keepers = ['m4A-301-1', 'm4A-301-2', 'm4A-500-1', 'm4A-500-2', 'm4A-712-1', 'm4A-712-2', 'm4A-736-1', 
           'm4A-736-2', 'm4A-744-1', 'm4A-744-2', 'm4A-755-1', 'm4A-755-2', 'm4A-757-1', 'm4A-757-2', 
           'm4A-766-1', 'm4A-766-2', 'm4A-mock3-16000', 'm4A-mock3-32000a', 'm4A-mock3-32000b', 'm4A-stds']
mock_samples = ['m4A-mock3-16000', 'm4A-mock3-32000a', 'm4A-mock3-32000b', 'm4A-stds']
samples = [item for item in keepers if item not in mock_samples]
#print samples
df2 = df.reindex(columns=keepers, index=natsorted(df.index))

#now drop OTUs where they are all zero
df3 = df2.loc[(df2 != 0).any(1)]
mock_otus = []
for x in df3.index:
    if x.startswith('mock'):
        if not x.startswith('mock3'): #mock3 is actually in some of the samples so can't use for index-bleed measurment
            mock_otus.append(x)
#now slice into two tables containing only mocks and calculate index-bleed
bleed_counts = df3.reindex(columns=samples, index=mock_otus)
bleed_sum = bleed_counts.values.sum()
real_counts = df3.reindex(columns=mock_samples, index=mock_otus)
real_sum = real_counts.values.sum()
index_bleed = int(bleed_sum) / float(real_sum) * 100
print "Index-bleed out of mock: %f%% (%i out of %i)" % (index_bleed, bleed_sum, real_sum)

#now try to to do the opposite, count number of times that samples move into mock
#first generate list of mock otus
sample_otus = []
for x in df3.index:
    if not x.startswith('mock'):
        sample_otus.append(x)
#slice first to samples, further trim list of only things found in mocks, would be spurious or chimeras
sampledf = df3.reindex(columns=samples, index=sample_otus)
sampledf = sampledf.loc[(sampledf != 0).any(1)]
cleaned_otus = sampledf.index
#print len(cleaned_otus), len(sample_otus)

#now using new list grab counts in mock
bleed_counts = df3.reindex(columns=mock_samples, index=cleaned_otus)
bleed_counts = bleed_counts.loc[(bleed_counts != 0).any(1)]
bleed_sum = bleed_counts.values.sum()
#print len(bleed_counts.index), bleed_sum

#finally go back to samples and get read counts for the ovlerlapping otus
real_counts = df3.reindex(columns=samples, index=bleed_counts.index)
real_sum = real_counts.values.sum()
index_bleed2 = int(bleed_sum) / float(real_sum) * 100
print "Index-bleed into mock: %f%% (%i out of %i)" % (index_bleed2, bleed_sum, real_sum)
print "Total index-bleed: %f%%" % (index_bleed + index_bleed2)

Index-bleed out of mock: 0.248094% (2377 out of 958106)
Index-bleed into mock: 0.014913% (403 out of 2702316)
Total index-bleed: 0.263007%


In [50]:
#need to measure index bleed of miseq data now, trying 0 mismatch miseqC dataset first
import pandas as pd
import numpy as np
from natsort import natsorted
import sys
#now load sorted "raw" OTU table into dataframe
df = pd.read_csv('miseq.final.txt', sep='\t', index_col=0)
#print list(df.columns.values)
#sys.exit(1)
#print len(df)
#okay, so 1367 OTUs in entire data set, lets trim this down to only run 3, as that one is labeled here with biomock
keepers = ['m4C-301-1', 'm4C-301-2', 'm4C-500-1', 'm4C-500-2', 'm4C-712-1', 'm4C-712-2', 
           'm4C-736-1', 'm4C-736-2', 'm4C-744-1', 'm4C-744-2', 'm4C-755-1', 'm4C-755-2', 
           'm4C-757-1', 'm4C-757-2', 'm4C-766-1', 'm4C-766-2', 
           'm4C-mock3-16000', 'm4C-mock3-32000a', 'm4C-mock3-32000b', 'm4C-stds']
mock_samples = ['m4C-mock3-16000', 'm4C-mock3-32000a', 'm4C-mock3-32000b', 'm4C-stds']
samples = [item for item in keepers if item not in mock_samples]
#print samples
df2 = df.reindex(columns=keepers, index=natsorted(df.index))

#now drop OTUs where they are all zero
df3 = df2.loc[(df2 != 0).any(1)]
mock_otus = []
for x in df3.index:
    if x.startswith('mock'):
        if not x.startswith('mock3'): #mock3 is actually in some of the samples so can't use for index-bleed measurment
            mock_otus.append(x)
#now slice into two tables containing only mocks and calculate index-bleed
bleed_counts = df3.reindex(columns=samples, index=mock_otus)
bleed_sum = bleed_counts.values.sum()
real_counts = df3.reindex(columns=mock_samples, index=mock_otus)
real_sum = real_counts.values.sum()
index_bleed = int(bleed_sum) / float(real_sum) * 100
print "Index-bleed out of mock: %f%% (%i out of %i)" % (index_bleed, bleed_sum, real_sum)

#now try to to do the opposite, count number of times that samples move into mock
#first generate list of mock otus
sample_otus = []
for x in df3.index:
    if not x.startswith('mock'):
        sample_otus.append(x)
#slice first to samples, further trim list of only things found in mocks, would be spurious or chimeras
sampledf = df3.reindex(columns=samples, index=sample_otus)
sampledf = sampledf.loc[(sampledf != 0).any(1)]
cleaned_otus = sampledf.index
#print len(cleaned_otus), len(sample_otus)

#now using new list grab counts in mock
bleed_counts = df3.reindex(columns=mock_samples, index=cleaned_otus)
bleed_counts = bleed_counts.loc[(bleed_counts != 0).any(1)]
bleed_sum = bleed_counts.values.sum()
#print len(bleed_counts.index), bleed_sum

#finally go back to samples and get read counts for the ovlerlapping otus
real_counts = df3.reindex(columns=samples, index=bleed_counts.index)
real_sum = real_counts.values.sum()
index_bleed2 = int(bleed_sum) / float(real_sum) * 100
print "Index-bleed into mock: %f%% (%i out of %i)" % (index_bleed2, bleed_sum, real_sum)
print "Total index-bleed: %f%%" % (index_bleed + index_bleed2)

Index-bleed out of mock: 0.216339% (1762 out of 814461)
Index-bleed into mock: 0.012101% (294 out of 2429463)
Total index-bleed: 0.228441%


In [92]:
#now measure index bleed using the biomock for m6
#need to measure index bleed of miseq data now, trying 0 mismatch miseqC dataset first
import pandas as pd
import numpy as np
from natsort import natsorted
import sys
#now load sorted "raw" OTU table into dataframe
df = pd.read_csv('miseq.final.txt', sep='\t', index_col=0)
#print list(df.columns.values)
#sys.exit(1)
#print len(df)
#okay, so 1367 OTUs in entire data set, lets trim this down to only run 3, as that one is labeled here with biomock
keepers = ['m6-301-1', 'm6-301-2', 'm6-500-1', 'm6-500-2', 'm6-712-1', 'm6-712-2', 
           'm6-736-1', 'm6-736-2', 'm6-744-1', 'm6-744-2', 'm6-755-1', 'm6-755-2', 
           'm6-757-1', 'm6-757-2', 'm6-766-1', 'm6-766-2', 
           'm6-mock3-16000', 'm6-mock3-32000a', 'm6-mock3-32000b', 'm6-stds', 'm6-SynMock']
mock_samples = ['m6-mock3-16000', 'm6-mock3-32000a', 'm6-mock3-32000b', 'm6-stds']
samples = [item for item in keepers if item not in mock_samples]
#print samples
df2 = df.reindex(columns=keepers, index=natsorted(df.index))

#now drop OTUs where they are all zero
df3 = df2.loc[(df2 != 0).any(1)]
mock_otus = []
for x in df3.index:
    if x.startswith('mock'):
        if not x.startswith('mock3'): #mock3 is actually in some of the samples so can't use for index-bleed measurment
            mock_otus.append(x)
#now slice into two tables containing only mocks and calculate index-bleed
bleed_counts = df3.reindex(columns=samples, index=mock_otus)
bleed_sum = bleed_counts.values.sum()
real_counts = df3.reindex(columns=mock_samples, index=mock_otus)
real_sum = real_counts.values.sum()
index_bleed = int(bleed_sum) / float(real_sum) * 100
print "Index-bleed out of mock: %f%% (%i out of %i)" % (index_bleed, bleed_sum, real_sum)

#now try to to do the opposite, count number of times that samples move into mock
#first generate list of mock otus
sample_otus = []
for x in df3.index:
    if not x.startswith('mock'):
        sample_otus.append(x)
#slice first to samples, further trim list of only things found in mocks, would be spurious or chimeras
sampledf = df3.reindex(columns=samples, index=sample_otus)
sampledf = sampledf.loc[(sampledf != 0).any(1)]
cleaned_otus = sampledf.index
#print len(cleaned_otus), len(sample_otus)

#now using new list grab counts in mock
bleed_counts = df3.reindex(columns=mock_samples, index=cleaned_otus)
bleed_counts = bleed_counts.loc[(bleed_counts != 0).any(1)]
bleed_sum = bleed_counts.values.sum()
#print len(bleed_counts.index), bleed_sum

#finally go back to samples and get read counts for the ovlerlapping otus
real_counts = df3.reindex(columns=samples, index=bleed_counts.index)
real_sum = real_counts.values.sum()
index_bleed2 = int(bleed_sum) / float(real_sum) * 100
print "Index-bleed into mock: %f%% (%i out of %i)" % (index_bleed2, bleed_sum, real_sum)
print "Total index-bleed: %f%%" % (index_bleed + index_bleed2)

Index-bleed out of mock: 0.244371% (6671 out of 2729864)
Index-bleed into mock: 0.032036% (3027 out of 9448644)
Total index-bleed: 0.276407%


In [52]:
#now measure index bleed using synmock data set in m6 run
#need to measure index bleed of miseq data now, trying 0 mismatch miseqC dataset first
import pandas as pd
import numpy as np
from natsort import natsorted
import sys
#now load sorted "raw" OTU table into dataframe
df = pd.read_csv('miseq_syn.final.txt', sep='\t', index_col=0)
#print list(df.columns.values)
#sys.exit(1)
#print len(df)
#okay, so 1367 OTUs in entire data set, lets trim this down to only run 3, as that one is labeled here with biomock
keepers = ['m6-301-1', 'm6-301-2', 'm6-500-1', 'm6-500-2', 'm6-712-1', 'm6-712-2', 
           'm6-736-1', 'm6-736-2', 'm6-744-1', 'm6-744-2', 'm6-755-1', 'm6-755-2', 
           'm6-757-1', 'm6-757-2', 'm6-766-1', 'm6-766-2', 
           'm6-mock3-16000', 'm6-mock3-32000a', 'm6-mock3-32000b', 'm6-stds', 'm6-SynMock']
mock_samples = ['m6-SynMock']
samples = [item for item in keepers if item not in mock_samples]
#print samples
df2 = df.reindex(columns=keepers, index=natsorted(df.index))

#now drop OTUs where they are all zero
df3 = df2.loc[(df2 != 0).any(1)]
mock_otus = []
for x in df3.index:
    if x.startswith('mock'):
        if not x.startswith('mock3'): #mock3 is actually in some of the samples so can't use for index-bleed measurment
            mock_otus.append(x)
#now slice into two tables containing only mocks and calculate index-bleed
bleed_counts = df3.reindex(columns=samples, index=mock_otus)
bleed_sum = bleed_counts.values.sum()
real_counts = df3.reindex(columns=mock_samples, index=mock_otus)
real_sum = real_counts.values.sum()
index_bleed = int(bleed_sum) / float(real_sum) * 100
print "Index-bleed out of mock: %f%% (%i out of %i)" % (index_bleed, bleed_sum, real_sum)

#now try to to do the opposite, count number of times that samples move into mock
#first generate list of mock otus
sample_otus = []
for x in df3.index:
    if not x.startswith('mock'):
        sample_otus.append(x)
#slice first to samples, further trim list of only things found in mocks, would be spurious or chimeras
sampledf = df3.reindex(columns=samples, index=sample_otus)
sampledf = sampledf.loc[(sampledf != 0).any(1)]
cleaned_otus = sampledf.index
#print len(cleaned_otus), len(sample_otus)

#now using new list grab counts in mock
bleed_counts = df3.reindex(columns=mock_samples, index=cleaned_otus)
bleed_counts = bleed_counts.loc[(bleed_counts != 0).any(1)]
bleed_sum = bleed_counts.values.sum()
#print len(bleed_counts.index), bleed_sum

#finally go back to samples and get read counts for the ovlerlapping otus
real_counts = df3.reindex(columns=samples, index=bleed_counts.index)
real_sum = real_counts.values.sum()
index_bleed2 = int(bleed_sum) / float(real_sum) * 100
print "Index-bleed into mock: %f%% (%i out of %i)" % (index_bleed2, bleed_sum, real_sum)
print "Total index-bleed: %f%%" % (index_bleed + index_bleed2)

Index-bleed out of mock: 0.032520% (340 out of 1045519)
Index-bleed into mock: 0.013675% (683 out of 4994664)
Total index-bleed: 0.046194%


In [78]:
#want to measure index bleed, also have 1 mismatch dataset, so generate OTU table for that dataset
!vsearch --fastq_filter m4a1m.demux.fq -fastaout m4a1m.demux.fa
!vsearch --usearch_global m4a1m.demux.fa --db miseq.cluster.otus.fa --id 0.97 --strand plus --otutabout miseq4a1m.otu_table.txt
!rm m4a1m.demux.fa

vsearch v2.4.3_osx_x86_64, 16.0GB RAM, 8 cores
https://github.com/torognes/vsearch

Reading input file 100%                                                                                                                                                                                                                                                                                                                                                                                                              
5946978 sequences kept (of which 0 truncated), 0 sequences discarded.
vsearch v2.4.3_osx_x86_64, 16.0GB RAM, 8 cores
https://github.com/torognes/vsearch

Reading file miseq.cluster.otus.fa 100%  
744596 nt in 2816 seqs, min 100, max 300, avg 264
Masking 100%  
Counting unique k-mers 100%  
Creating index of unique k-mers 100%  
Searching 100%                                                                                                                                                      

In [79]:
#run through filtering so it is sorted etc
!amptk filter -i miseq4a1m.otu_table.txt -f miseq.cluster.otus.fa -p 0 -b m4a-1mm-stds -m mock2 --keep_mock -o miseq4a1m

-------------------------------------------------------
[11:29:52 PM]: OS: MacOSX 10.12.5, 8 cores, ~ 16 GB RAM. Python: 2.7.12
[11:29:52 PM]: AMPtk v.0.10.0, USEARCH v9.2.64, VSEARCH v2.4.3
[11:29:52 PM]: Loading OTU table: miseq4a1m.otu_table.txt
[11:29:52 PM]: OTU table contains 2399 OTUs
[11:29:52 PM]: Mapping OTUs to Mock Community (USEARCH)
[11:29:52 PM]: Sorting OTU table naturally
[11:29:52 PM]: Removing OTUs according to --min_reads_otu: (OTUs with less than 2 reads from all samples)
[11:29:52 PM]: Normalizing OTU table to number of reads per sample
[11:29:52 PM]: Index bleed, mock into samples: 75.585340%.  Index bleed, samples into mock: 0.389027%.
[11:29:52 PM]: Overwriting auto detect index-bleed, setting to 0.000000%
[11:29:54 PM]: Filtering OTU table down to 1989 OTUs
[11:29:54 PM]: Filtering valid OTUs
-------------------------------------------------------
OTU Table filtering finished
-------------------------------------------------------
OTU Table Stats:      miseq4a

In [81]:
#need to measure index bleed of miseq data now, trying 0 mismatch miseqA dataset
import pandas as pd
import numpy as np
from natsort import natsorted
import sys
#now load sorted "raw" OTU table into dataframe
df = pd.read_csv('miseq4a1m.sorted.txt', sep='\t', index_col=0)
#print list(df.columns.values)
#sys.exit(1)
#print len(df)
#okay, so 1367 OTUs in entire data set, lets trim this down to only run 3, as that one is labeled here with biomock
keepers = ['m4a-1mm-301-1','m4a-1mm-301-2','m4a-1mm-500-1','m4a-1mm-500-2',
           'm4a-1mm-712-1','m4a-1mm-712-2','m4a-1mm-736-1','m4a-1mm-736-2',
           'm4a-1mm-744-1','m4a-1mm-744-2','m4a-1mm-755-1','m4a-1mm-755-2',
           'm4a-1mm-757-1','m4a-1mm-757-2','m4a-1mm-766-1','m4a-1mm-766-2',
           'm4a-1mm-mock3-16000','m4a-1mm-mock3-32000a','m4a-1mm-mock3-32000b','m4a-1mm-stds']
mock_samples = ['m4a-1mm-mock3-16000','m4a-1mm-mock3-32000a','m4a-1mm-mock3-32000b','m4a-1mm-stds']
samples = [item for item in keepers if item not in mock_samples]
#print samples
df2 = df.reindex(columns=keepers, index=natsorted(df.index))

#now drop OTUs where they are all zero
df3 = df2.loc[(df2 != 0).any(1)]
mock_otus = []
for x in df3.index:
    if x.startswith('mock'):
        if not x.startswith('mock3'): #mock3 is actually in some of the samples so can't use for index-bleed measurment
            mock_otus.append(x)
#now slice into two tables containing only mocks and calculate index-bleed
bleed_counts = df3.reindex(columns=samples, index=mock_otus)
bleed_sum = bleed_counts.values.sum()
real_counts = df3.reindex(columns=mock_samples, index=mock_otus)
real_sum = real_counts.values.sum()
index_bleed = int(bleed_sum) / float(real_sum) * 100
print "Index-bleed out of mock: %f%% (%i out of %i)" % (index_bleed, bleed_sum, real_sum)

#now try to to do the opposite, count number of times that samples move into mock
#first generate list of mock otus
sample_otus = []
for x in df3.index:
    if not x.startswith('mock'):
        sample_otus.append(x)
#slice first to samples, further trim list of only things found in mocks, would be spurious or chimeras
sampledf = df3.reindex(columns=samples, index=sample_otus)
sampledf = sampledf.loc[(sampledf != 0).any(1)]
cleaned_otus = sampledf.index
#print len(cleaned_otus), len(sample_otus)

#now using new list grab counts in mock
bleed_counts = df3.reindex(columns=mock_samples, index=cleaned_otus)
bleed_counts = bleed_counts.loc[(bleed_counts != 0).any(1)]
bleed_sum = bleed_counts.values.sum()
#print len(bleed_counts.index), bleed_sum

#finally go back to samples and get read counts for the ovlerlapping otus
real_counts = df3.reindex(columns=samples, index=bleed_counts.index)
real_sum = real_counts.values.sum()
index_bleed2 = int(bleed_sum) / float(real_sum) * 100
print "Index-bleed into mock: %f%% (%i out of %i)" % (index_bleed2, bleed_sum, real_sum)
print "Total index-bleed: %f%%" % (index_bleed + index_bleed2)

Index-bleed out of mock: 0.323377% (3206 out of 991411)
Index-bleed into mock: 0.033259% (1038 out of 3120998)
Total index-bleed: 0.356636%


In [82]:
#measure index bleed of miseq61 mismatch dataset, so generate OTU table for that dataset
!vsearch --fastq_filter m61mm.demux.fq -fastaout m61mm.demux.fa
!vsearch --usearch_global m61mm.demux.fa --db miseq.cluster.otus.fa --id 0.97 --strand plus --otutabout miseq61mm.otu_table.txt
!rm m61mm.demux.fa
!amptk filter -i miseq61mm.otu_table.txt -f miseq.cluster.otus.fa -p 0 -b m6-stds -m mock2 --keep_mock -o miseq61mm

vsearch v2.4.3_osx_x86_64, 16.0GB RAM, 8 cores
https://github.com/torognes/vsearch

Reading input file 100%                                                                                                                                                                                                                                                                                                                                                                                                              
19874272 sequences kept (of which 0 truncated), 0 sequences discarded.
vsearch v2.4.3_osx_x86_64, 16.0GB RAM, 8 cores
https://github.com/torognes/vsearch

Reading file miseq.cluster.otus.fa 100%  
744596 nt in 2816 seqs, min 100, max 300, avg 264
Masking 100%   
Counting unique k-mers 100%  
Creating index of unique k-mers 100%  
Searching 100%                                                                                                                                                    

In [85]:
#now measure index bleed using synmock data set in m6 run
#need to measure index bleed of miseq data now, trying 0 mismatch miseqC dataset first
import pandas as pd
import numpy as np
from natsort import natsorted
import sys
#now load sorted "raw" OTU table into dataframe
df = pd.read_csv('miseq61mm.final.txt', sep='\t', index_col=0)
#print list(df.columns.values)
#sys.exit(1)
#print len(df)
#okay, so 1367 OTUs in entire data set, lets trim this down to only run 3, as that one is labeled here with biomock
keepers = ['m6-301-1', 'm6-301-2', 'm6-500-1', 'm6-500-2', 'm6-712-1', 'm6-712-2', 
           'm6-736-1', 'm6-736-2', 'm6-744-1', 'm6-744-2', 'm6-755-1', 'm6-755-2', 
           'm6-757-1', 'm6-757-2', 'm6-766-1', 'm6-766-2','m6-mock3-16000', 'm6-mock3-32000a', 'm6-mock3-32000b', 'm6-stds', 
           'm6-SynMock']
mock_samples = ['m6-mock3-16000', 'm6-mock3-32000a', 'm6-mock3-32000b', 'm6-stds']
samples = [item for item in keepers if item not in mock_samples]
#print samples
df2 = df.reindex(columns=keepers, index=natsorted(df.index))

#now drop OTUs where they are all zero
df3 = df2.loc[(df2 != 0).any(1)]
mock_otus = []
for x in df3.index:
    if x.startswith('mock'):
        if not x.startswith('mock3'): #mock3 is actually in some of the samples so can't use for index-bleed measurment
            mock_otus.append(x)
#now slice into two tables containing only mocks and calculate index-bleed
bleed_counts = df3.reindex(columns=samples, index=mock_otus)
bleed_sum = bleed_counts.values.sum()
real_counts = df3.reindex(columns=mock_samples, index=mock_otus)
real_sum = real_counts.values.sum()
index_bleed = int(bleed_sum) / float(real_sum) * 100
print "Index-bleed out of mock: %f%% (%i out of %i)" % (index_bleed, bleed_sum, real_sum)

#now try to to do the opposite, count number of times that samples move into mock
#first generate list of mock otus
sample_otus = []
for x in df3.index:
    if not x.startswith('mock'):
        sample_otus.append(x)
#slice first to samples, further trim list of only things found in mocks, would be spurious or chimeras
sampledf = df3.reindex(columns=samples, index=sample_otus)
sampledf = sampledf.loc[(sampledf != 0).any(1)]
cleaned_otus = sampledf.index
#print len(cleaned_otus), len(sample_otus)

#now using new list grab counts in mock
bleed_counts = df3.reindex(columns=mock_samples, index=cleaned_otus)
bleed_counts = bleed_counts.loc[(bleed_counts != 0).any(1)]
bleed_sum = bleed_counts.values.sum()
#print len(bleed_counts.index), bleed_sum

#finally go back to samples and get read counts for the ovlerlapping otus
real_counts = df3.reindex(columns=samples, index=bleed_counts.index)
real_sum = real_counts.values.sum()
index_bleed2 = int(bleed_sum) / float(real_sum) * 100
print "Index-bleed into mock: %f%% (%i out of %i)" % (index_bleed2, bleed_sum, real_sum)
print "Total index-bleed: %f%%" % (index_bleed + index_bleed2)

Index-bleed out of mock: 0.302544% (8778 out of 2901396)
Index-bleed into mock: 0.035714% (4094 out of 11463401)
Total index-bleed: 0.338258%


In [86]:
#now we want to measure index bleed using synmock
!amptk filter -i miseq61mm.otu_table.txt -f miseq.cluster.otus.fa -p 0 -b m6-SynMock -m synmock --keep_mock -o miseq61mm_syn

-------------------------------------------------------
[07:58:17 AM]: OS: MacOSX 10.12.5, 8 cores, ~ 16 GB RAM. Python: 2.7.12
[07:58:17 AM]: AMPtk v.0.10.0, USEARCH v9.2.64, VSEARCH v2.4.3
[07:58:17 AM]: Loading OTU table: miseq61mm.otu_table.txt
[07:58:17 AM]: OTU table contains 2745 OTUs
[07:58:17 AM]: Mapping OTUs to Mock Community (USEARCH)
[07:58:17 AM]: Sorting OTU table naturally
[07:58:17 AM]: Removing OTUs according to --min_reads_otu: (OTUs with less than 2 reads from all samples)
[07:58:17 AM]: Normalizing OTU table to number of reads per sample
[07:58:17 AM]: Index bleed, mock into samples: 0.074052%.  Index bleed, samples into mock: 0.133015%.
[07:58:17 AM]: Overwriting auto detect index-bleed, setting to 0.000000%
[07:58:18 AM]: Filtering OTU table down to 2050 OTUs
[07:58:18 AM]: Filtering valid OTUs
-------------------------------------------------------
OTU Table filtering finished
-------------------------------------------------------
OTU Table Stats:      miseq61m

In [89]:
#now measure index bleed using synmock data set in m6 run
#need to measure index bleed of miseq data now, trying 0 mismatch miseqC dataset first
import pandas as pd
import numpy as np
from natsort import natsorted
import sys
#now load sorted "raw" OTU table into dataframe
df = pd.read_csv('miseq61mm_syn.final.txt', sep='\t', index_col=0)
#print list(df.columns.values)
#sys.exit(1)
#print len(df)
#okay, so 1367 OTUs in entire data set, lets trim this down to only run 3, as that one is labeled here with biomock
keepers = ['m6-301-1', 'm6-301-2', 'm6-500-1', 'm6-500-2', 'm6-712-1', 'm6-712-2', 
           'm6-736-1', 'm6-736-2', 'm6-744-1', 'm6-744-2', 'm6-755-1', 'm6-755-2', 
           'm6-757-1', 'm6-757-2', 'm6-766-1', 'm6-766-2','m6-mock3-16000', 'm6-mock3-32000a', 'm6-mock3-32000b', 'm6-stds', 
           'm6-SynMock']
mock_samples = ['m6-SynMock']
samples = [item for item in keepers if item not in mock_samples]
#print samples
df2 = df.reindex(columns=keepers, index=natsorted(df.index))

#now drop OTUs where they are all zero
df3 = df2.loc[(df2 != 0).any(1)]
mock_otus = []
for x in df3.index:
    if x.startswith('mock'):
        mock_otus.append(x)
#now slice into two tables containing only mocks and calculate index-bleed
bleed_counts = df3.reindex(columns=samples, index=mock_otus)
bleed_sum = bleed_counts.values.sum()
real_counts = df3.reindex(columns=mock_samples, index=mock_otus)
real_sum = real_counts.values.sum()
index_bleed = int(bleed_sum) / float(real_sum) * 100
print "Index-bleed out of mock: %f%% (%i out of %i)" % (index_bleed, bleed_sum, real_sum)

#now try to to do the opposite, count number of times that samples move into mock
#first generate list of mock otus
sample_otus = []
for x in df3.index:
    if not x.startswith('mock'):
        sample_otus.append(x)
#slice first to samples, further trim list of only things found in mocks, would be spurious or chimeras
sampledf = df3.reindex(columns=samples, index=sample_otus)
sampledf = sampledf.loc[(sampledf != 0).any(1)]
cleaned_otus = sampledf.index
#print len(cleaned_otus), len(sample_otus)

#now using new list grab counts in mock
bleed_counts = df3.reindex(columns=mock_samples, index=cleaned_otus)
bleed_counts = bleed_counts.loc[(bleed_counts != 0).any(1)]
bleed_sum = bleed_counts.values.sum()
#print len(bleed_counts.index), bleed_sum

#finally go back to samples and get read counts for the ovlerlapping otus
real_counts = df3.reindex(columns=samples, index=bleed_counts.index)
real_sum = real_counts.values.sum()
index_bleed2 = int(bleed_sum) / float(real_sum) * 100
print "Index-bleed into mock: %f%% (%i out of %i)" % (index_bleed2, bleed_sum, real_sum)
print "Total index-bleed: %f%%" % (index_bleed + index_bleed2)

Index-bleed out of mock: 0.054651% (603 out of 1103361)
Index-bleed into mock: 0.017822% (1238 out of 6946547)
Total index-bleed: 0.072473%


In [2]:
#need to measure index bleed of miseqB dataset, this one slightly differnt as can only measure index out of mock
import pandas as pd
import numpy as np
from natsort import natsorted
import sys
#now load sorted "raw" OTU table into dataframe
df = pd.read_csv('miseq.final.txt', sep='\t', index_col=0)
#print list(df.columns.values)
#sys.exit(1)
#print len(df)
#okay, so 1367 OTUs in entire data set, lets trim this down to only run 3, as that one is labeled here with biomock
keepers = ['m4b-744-1', 'm4b-744-2', 'm4b-766-1', 'm4b-766-2']
mock_samples = ['m4C-mock3-16000', 'm4C-mock3-32000a', 'm4C-mock3-32000b', 'm4C-stds']
samples = [item for item in keepers if item not in mock_samples]
#print samples
df2 = df.reindex(columns=keepers, index=natsorted(df.index))

#now drop OTUs where they are all zero
df3 = df2.loc[(df2 != 0).any(1)]
mock_otus = []
for x in df3.index: #get index from total table
    if x.startswith('mock'):
        if not x.startswith('mock3'): #mock3 is actually in some of the samples so can't use for index-bleed measurment
            mock_otus.append(x)
if len(mock_otus) < 1: #this means that no read counts from mock members were found, so index-bleed is 0
    print "Index-bleed out of mock: 0.0%"
    sys.exit(1)
#now slice into two tables containing only mocks and calculate index-bleed
bleed_counts = df3.reindex(columns=samples, index=mock_otus)
bleed_sum = bleed_counts.values.sum()
real_counts = df3.reindex(columns=mock_samples, index=mock_otus)
real_sum = real_counts.values.sum()
index_bleed = int(bleed_sum) / float(real_sum) * 100
print "Index-bleed out of mock: %f%% (%i out of %i)" % (index_bleed, bleed_sum, real_sum)

Index-bleed out of mock: 0.0%


SystemExit: 1

/Users/jon/miniconda2/lib/python2.7/site-packages/IPython/core/interactiveshell.py:2889: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.
  warn("To exit: use 'exit', 'quit', or Ctrl-D.", stacklevel=1)


| Run    | BC mismatch |  Mock   | Index-bleed out | Index-bleed in | Total index bleed |
| ------ | :---------: | :-----: | :-------------: | :------------: | :---------------: |
| Ion3   |      0      | biomock |     0.025%      |     0.018%     |      0.043%       |
| Ion5   |      0      | synmock |     0.041%      |     0.028%     |      0.069%       |
| MiseqA |      0      | biomock |     0.248%      |     0.015%     |      0.263%       |
| MiseqA |      1      | biomock |     0.323%      |     0.033%     |      0.356%       |
| MiSeqC |      0      | biomock |     0.216%      |     0.012%     |      0.228%       |
| Miseq6 |      0      | biomock |     0.244%      |     0.032%     |      0.276%       |
| Miseq6 |      0      | synmock |     0.033%      |     0.014%     |      0.046%       |
| Miseq6 |      1      | biomock |     0.303%      |     0.035%     |      0.338%       |
| Miseq6 |      1      | synmock |     0.055%      |     0.018%     |      0.073%       |



These data suggest that measuring index-bleed with biomock is over-estimating the cross-over.  These data also suggest that one can limit some index bleed by demuxing the data using a BC (index) mismatch of 0 instead of the default of 1 on MiSeq runs.  On average, there is a 10% reduction in the number of reads retained if BC mismatch is set to 0 versus 1 (default) on the Miseq6 run, however there was an increase in 63% in total index-bleed (0.046% to 0.073%), suggesting that a larger percentage of the reads containing BC mismatches are sample mis-classified.

#### Index Bleeds to use for filtering, round up to nearest 10th percentage. will use max read count for threshold for each
Ion3: 0.001

Ion5: 0.001

MiSeqA: 0.003

MiSeqC: 0.003

MiSeq6: 0.001

In [52]:
#generate OTU table for each of the runs using the combined OTUs from either miseq or ion
#now slice the OTU table for each of the runs, Ion3, Miseq4A, Miseq4C, Ion5
#load original table, not the name changed version for normal processing
import pandas as pd
from natsort import natsorted

miseqdf = pd.read_csv('miseq.otu_table.txt', sep='\t', index_col=0)
iondf = pd.read_csv('ion.otu_table.txt', sep='\t', index_col=0)

ion3 = ['m3-301-1', 'm3-301-2', 'm3-500-1', 'm3-500-2', 'm3-712-1', 'm3-712-2', 
        'm3-736-1', 'm3-736-2', 'm3-744-1', 'm3-744-2', 'm3-755-1',
        'm3-757-2', 'm3-766-1', 'm3-766-2', 'm3-mock3-16000', 'm3-mock3-3200a', 'm3-mock3-32000b', 'm3-stds']
ion3df = iondf.reindex(columns=ion3, index=natsorted(iondf.index))
ion3df.to_csv('ion3.otu_table.txt', sep='\t')
miseq4a = ['m4A-301-1', 'm4A-301-2', 'm4A-500-1', 'm4A-500-2', 'm4A-712-1', 
           'm4A-712-2', 'm4A-736-1', 'm4A-736-2', 'm4A-744-1', 'm4A-744-2', 
           'm4A-755-1', 'm4A-755-2', 'm4A-757-1', 'm4A-757-2', 'm4A-766-1', 'm4A-766-2',
           'm4A-mock3-16000', 'm4A-mock3-32000a', 'm4A-mock3-32000b', 'm4A-stds']
miseq4adf = miseqdf.reindex(columns=miseq4a, index=natsorted(miseqdf.index))
miseq4adf.to_csv('miseq4a.otu_table.txt', sep='\t')
miseq4c = ['m4C-301-1', 'm4C-301-2', 'm4C-500-1', 'm4C-500-2', 'm4C-712-1', 
           'm4C-712-2', 'm4C-736-1', 'm4C-736-2', 'm4C-744-1', 'm4C-744-2', 
           'm4C-755-1', 'm4C-755-2', 'm4C-757-1', 'm4C-757-2', 'm4C-766-1', 'm4C-766-2',
          'm4C-mock3-16000', 'm4C-mock3-32000a', 'm4C-mock3-32000b', 'm4C-stds']
miseq4cdf = miseqdf.reindex(columns=miseq4c, index=natsorted(miseqdf.index))
miseq4cdf.to_csv('miseq4c.otu_table.txt', sep='\t')
miseq6 = ['m6-301-1', 'm6-301-2', 'm6-500-1', 'm6-500-2', 'm6-712-1', 
           'm6-712-2', 'm6-736-1', 'm6-736-2', 'm6-744-1', 'm6-744-2', 
           'm6-755-1', 'm6-755-2', 'm6-757-1', 'm6-757-2', 'm6-766-1', 'm6-766-2', 
          'm6-mock3-16000', 'm6-mock3-32000a', 'm6-mock3-32000b', 'm6-stds','m6-SynMock']
miseq6df = miseqdf.reindex(columns=miseq6, index=natsorted(miseqdf.index))
miseq6df.to_csv('miseq6.otu_table.txt', sep='\t')
ion5 = ['m5-301-1', 'm5-301-2', 'm5-500-1', 'm5-500-2', 'm5-712-1', 
        'm5-712-2', 'm5-736-1', 'm5-736-2', 'm5-744-1', 'm5-744-2', 
        'm5-755-1', 'm5-755-2', 'm5-757-1', 'm5-757-2', 'm5-766-1', 'm5-766-2',
        'm5-SynMock1-8000', 'm5-SynMock1-mock3.0', 'm5-SynMock1-mock3.5', 'm5-SynMock1.736-1.736-2',
        'm5-SynMock2-8000', 'm5-SynMock2.757-1.757-2', 'm5-mock3.0-16000', 'm5-mock3.5-16000']
ion5df = iondf.reindex(columns=ion5, index=natsorted(iondf.index))
ion5df.to_csv('ion5.otu_table.txt', sep='\t')

In [56]:
#now filter these datasets for raw and filtered (no subtraction, index-bleed based on calculated values above)
!amptk filter -f miseq.cluster.otus.fa -i miseq4a.otu_table.txt -o miseq4a -b m4A-mock3-16000 -m mock3 -p 0.003 -c in -d m4A-mock3-32000a m4A-mock3-32000b m4A-stds 
!amptk filter -f miseq.cluster.otus.fa -i miseq4c.otu_table.txt -o miseq4c -b m4C-mock3-16000 -m mock3 -p 0.003 -c in -d m4C-mock3-32000a m4C-mock3-32000 m4C-stds
!amptk filter -f miseq.cluster.otus.fa -i miseq6.otu_table.txt -o miseq6 -b m6-SynMock -m synmock -p 0.001 -d m6-mock3-16000 m6-mock3-32000a m6-mock3-32000b m6-stds
!amptk filter -f ion.cluster.otus.fa -i ion3.otu_table.txt -o ion3 -b m3-mock3-16000 -m mock3 -p 0.001 -c in -d m3-mock3-3200a m3-mock3-32000b m3-stds
!amptk filter -f ion.cluster.otus.fa -i ion5.otu_table.txt -o ion5 -b m5-SynMock2-8000 -m synmock -p 0.001 -d m5-SynMock1-8000 m5-SynMock1-mock3.0 m5-SynMock1-mock3.5 m5-SynMock1.736-1.736-2 m5-SynMock2.757-1.757-2 m5-mock3.0-16000 m5-mock3.5-16000

-------------------------------------------------------
[02:34:09 PM]: OS: MacOSX 10.12.5, 8 cores, ~ 16 GB RAM. Python: 2.7.12
[02:34:09 PM]: AMPtk v.0.10.0, USEARCH v9.2.64, VSEARCH v2.4.3
[02:34:09 PM]: Loading OTU table: miseq4a.otu_table.txt
[02:34:09 PM]: OTU table contains 2816 OTUs
[02:34:09 PM]: Mapping OTUs to Mock Community (USEARCH)
[02:34:09 PM]: Sorting OTU table naturally
[02:34:09 PM]: Removing OTUs according to --min_reads_otu: (OTUs with less than 2 reads from all samples)
[02:34:09 PM]: Normalizing OTU table to number of reads per sample
[02:34:09 PM]: Index bleed, samples into mock: 0.059001%.
[02:34:09 PM]: Overwriting auto detect index-bleed, setting to 0.300000%
[02:34:09 PM]: Dropping 3 samples from table: m4A-mock3-32000a, m4A-mock3-32000b, m4A-stds
[02:34:10 PM]: Filtering OTU table down to 1937 OTUs
[02:34:10 PM]: Filtering valid OTUs
-------------------------------------------------------
OTU Table filtering finished
-----------------------------------------

In [39]:
#try to filter the miseq6 dataset based on the biomock index bleed as well.
#first create OTU table
import pandas as pd
from natsort import natsorted
miseqdf = pd.read_csv('miseq.otu_table.txt', sep='\t', index_col=0)
miseq6 = ['m6-301-1', 'm6-301-2', 'm6-500-1', 'm6-500-2', 'm6-712-1', 
           'm6-712-2', 'm6-736-1', 'm6-736-2', 'm6-744-1', 'm6-744-2', 
           'm6-755-1', 'm6-755-2', 'm6-757-1', 'm6-757-2', 'm6-766-1', 'm6-766-2', 'm6-mock3-16000']
miseq6df = miseqdf.reindex(columns=miseq6, index=natsorted(miseqdf.index))
miseq6df.to_csv('miseq6_biomock.otu_table.txt', sep='\t')
!amptk filter -f miseq.cluster.otus.fa -i miseq6_biomock.otu_table.txt -o miseq6_biomock -b m6-mock3-16000 -m mock3 -p 0.0028
!amptk taxonomy -f miseq6_biomock.filtered.otus.fa -i miseq6_biomock.final.txt -o miseq6_biomock -d ITS2 -t miseq.taxonomy.txt -m mapping/miseq6.mapping_file.txt

-------------------------------------------------------
[10:45:47 AM]: OS: MacOSX 10.12.5, 8 cores, ~ 16 GB RAM. Python: 2.7.12
[10:45:47 AM]: AMPtk v.0.10.0, USEARCH v9.2.64, VSEARCH v2.4.3
[10:45:47 AM]: Loading OTU table: miseq6_biomock.otu_table.txt
[10:45:47 AM]: OTU table contains 2816 OTUs
[10:45:47 AM]: Mapping OTUs to Mock Community (USEARCH)
[10:45:47 AM]: Sorting OTU table naturally
[10:45:48 AM]: Removing OTUs according to --min_reads_otu: (OTUs with less than 2 reads from all samples)
[10:45:48 AM]: Normalizing OTU table to number of reads per sample
[10:45:48 AM]: Index bleed, mock into samples: 7.510277%.  Index bleed, samples into mock: 0.094006%.
[10:45:48 AM]: Overwriting auto detect index-bleed, setting to 0.280000%
[10:45:48 AM]: Filtering OTU table down to 2015 OTUs
[10:45:48 AM]: Filtering valid OTUs
-------------------------------------------------------
OTU Table filtering finished
-------------------------------------------------------
OTU Table Stats:      mis

In [65]:
#now generate the raw OTU table from the sorted (original output)
import pandas as pd
from natsort import natsorted
import subprocess
#pull out header for indexing later
base = 'ion3'
platform = 'ion'
    
#now load sorted "raw" OTU table into dataframe
sort = pd.read_csv(base+'.sorted.txt', sep='\t', index_col=0)

#now we want to drop last column
header = ['m3-301-1', 'm3-301-2', 'm3-500-1', 'm3-500-2', 'm3-712-1', 'm3-712-2', 
        'm3-736-1', 'm3-736-2', 'm3-744-1', 'm3-744-2', 'm3-755-1',
        'm3-757-2', 'm3-766-1', 'm3-766-2']
cleaned = sort.reindex(columns=header, index=natsorted(sort.index))

#now drop OTUs where they are all zero
fs = cleaned.sum(axis=0)
filtered = pd.DataFrame(cleaned, columns=fs.index)
filt2 = filtered.loc[(filtered != 0).any(1)]
print len(sort), len(sort.columns.values), len(filt2), len(filt2.columns.values)
filt2.to_csv(base+'.sorted.nomock.txt', sep='\t')

#now convert to biom
cmd = ['amptk', 'taxonomy', '-f', platform+'.cluster.otus.fa', '-o', base+'_raw', '-i', base+'.sorted.nomock.txt', '-d', 'ITS2','-m','mapping/'+base+'.mapping_file.txt', '-t', platform+'.taxonomy.txt']
subprocess.call(cmd)

1310 15 1158 14


0

In [61]:
#now generate the raw OTU table from the sorted (original output)
import pandas as pd
from natsort import natsorted
import subprocess
#pull out header for indexing later
base = 'ion5'
platform = 'ion'
    
#now load sorted "raw" OTU table into dataframe
sort = pd.read_csv(base+'.sorted.txt', sep='\t', index_col=0)

#now we want to drop last column
header = ['m5-301-1', 'm5-301-2', 'm5-500-1', 'm5-500-2', 'm5-712-1', 
        'm5-712-2', 'm5-736-1', 'm5-736-2', 'm5-744-1', 'm5-744-2', 
        'm5-755-1', 'm5-755-2', 'm5-757-1', 'm5-757-2', 'm5-766-1', 'm5-766-2']
cleaned = sort.reindex(columns=header, index=natsorted(sort.index))

#now drop OTUs where they are all zero
fs = cleaned.sum(axis=0)
filtered = pd.DataFrame(cleaned, columns=fs.index)
filt2 = filtered.loc[(filtered != 0).any(1)]
print len(sort), len(sort.columns.values), len(filt2), len(filt2.columns.values)
filt2.to_csv(base+'.sorted.nomock.txt', sep='\t')

#now convert to biom
cmd = ['amptk', 'taxonomy', '-f', platform+'.cluster.otus.fa', '-o', base+'_raw', '-i', base+'.sorted.nomock.txt', '-d', 'ITS2','-m','mapping/'+base+'.mapping_file.txt', '-t', platform+'.taxonomy.txt']
subprocess.call(cmd)

1310 17 1179 16


0

In [63]:
#now generate the raw OTU table from the sorted (original output)
import pandas as pd
from natsort import natsorted
import subprocess
#pull out header for indexing later
base = 'miseq4a'
platform = 'miseq'
    
#now load sorted "raw" OTU table into dataframe
sort = pd.read_csv(base+'.sorted.txt', sep='\t', index_col=0)

#now we want to drop last column
header = ['m4A-301-1', 'm4A-301-2', 'm4A-500-1', 'm4A-500-2', 'm4A-712-1', 
           'm4A-712-2', 'm4A-736-1', 'm4A-736-2', 'm4A-744-1', 'm4A-744-2', 
           'm4A-755-1', 'm4A-755-2', 'm4A-757-1', 'm4A-757-2', 'm4A-766-1', 'm4A-766-2']
cleaned = sort.reindex(columns=header, index=natsorted(sort.index))

#now drop OTUs where they are all zero
fs = cleaned.sum(axis=0)
filtered = pd.DataFrame(cleaned, columns=fs.index)
filt2 = filtered.loc[(filtered != 0).any(1)]
print len(sort), len(sort.columns.values), len(filt2), len(filt2.columns.values)
filt2.to_csv(base+'.sorted.nomock.txt', sep='\t')

#now convert to biom
cmd = ['amptk', 'taxonomy', '-f', platform+'.cluster.otus.fa', '-o', base+'_raw', '-i', base+'.sorted.nomock.txt', '-d', 'ITS2','-m','mapping/'+base+'.mapping_file.txt', '-t', platform+'.taxonomy.txt']
subprocess.call(cmd)

2816 17 2368 16


0

In [62]:
#now generate the raw OTU table from the sorted (original output)
import pandas as pd
from natsort import natsorted
import subprocess
#pull out header for indexing later
base = 'miseq4c'
platform = 'miseq'
    
#now load sorted "raw" OTU table into dataframe
sort = pd.read_csv(base+'.sorted.txt', sep='\t', index_col=0)

#now we want to drop last column
header = ['m4C-301-1', 'm4C-301-2', 'm4C-500-1', 'm4C-500-2', 'm4C-712-1', 
           'm4C-712-2', 'm4C-736-1', 'm4C-736-2', 'm4C-744-1', 'm4C-744-2', 
           'm4C-755-1', 'm4C-755-2', 'm4C-757-1', 'm4C-757-2', 'm4C-766-1', 'm4C-766-2']
cleaned = sort.reindex(columns=header, index=natsorted(sort.index))

#now drop OTUs where they are all zero
fs = cleaned.sum(axis=0)
filtered = pd.DataFrame(cleaned, columns=fs.index)
filt2 = filtered.loc[(filtered != 0).any(1)]
print len(sort), len(sort.columns.values), len(filt2), len(filt2.columns.values)
filt2.to_csv(base+'.sorted.nomock.txt', sep='\t')

#now convert to biom
cmd = ['amptk', 'taxonomy', '-f', platform+'.cluster.otus.fa', '-o', base+'_raw', '-i', base+'.sorted.nomock.txt', '-d', 'ITS2','-m','mapping/'+base+'.mapping_file.txt', '-t', platform+'.taxonomy.txt']
subprocess.call(cmd)

2816 17 2271 16


0

In [64]:
#now generate the raw OTU table from the sorted (original output)
import pandas as pd
from natsort import natsorted
import subprocess
#pull out header for indexing later
base = 'miseq6'
platform = 'miseq'
    
#now load sorted "raw" OTU table into dataframe
sort = pd.read_csv(base+'.sorted.txt', sep='\t', index_col=0)

#now we want to drop last column
header = ['m6-301-1', 'm6-301-2', 'm6-500-1', 'm6-500-2', 'm6-712-1', 
           'm6-712-2', 'm6-736-1', 'm6-736-2', 'm6-744-1', 'm6-744-2', 
           'm6-755-1', 'm6-755-2', 'm6-757-1', 'm6-757-2', 'm6-766-1', 'm6-766-2']
cleaned = sort.reindex(columns=header, index=natsorted(sort.index))

#now drop OTUs where they are all zero
fs = cleaned.sum(axis=0)
filtered = pd.DataFrame(cleaned, columns=fs.index)
filt2 = filtered.loc[(filtered != 0).any(1)]
print len(sort), len(sort.columns.values), len(filt2), len(filt2.columns.values)
filt2.to_csv(base+'.sorted.nomock.txt', sep='\t')

#now convert to biom
cmd = ['amptk', 'taxonomy', '-f', platform+'.cluster.otus.fa', '-o', base+'_raw', '-i', base+'.sorted.nomock.txt', '-d', 'ITS2','-m','mapping/'+base+'.mapping_file.txt', '-t', platform+'.taxonomy.txt']
subprocess.call(cmd)

2816 17 2699 16


0

In [57]:
#run taxonomy command on all output
!amptk taxonomy -f ion3.filtered.otus.fa -i ion3.final.txt -o ion3 -d ITS2 -t ion.taxonomy.txt -m mapping/ion3.mapping_file.txt
!amptk taxonomy -f ion5.filtered.otus.fa -i ion5.final.txt -o ion5 -d ITS2 -t ion.taxonomy.txt -m mapping/ion5.mapping_file.txt
!amptk taxonomy -f miseq4a.filtered.otus.fa -i miseq4a.final.txt -o miseq4a -d ITS2 -t miseq.taxonomy.txt -m mapping/miseq4a.mapping_file.txt
!amptk taxonomy -f miseq4c.filtered.otus.fa -i miseq4c.final.txt -o miseq4c -d ITS2 -t miseq.taxonomy.txt -m mapping/miseq4c.mapping_file.txt
!amptk taxonomy -f miseq6.filtered.otus.fa -i miseq6.final.txt -o miseq6 -d ITS2 -t miseq.taxonomy.txt -m mapping/miseq6.mapping_file.txt

-------------------------------------------------------
[02:35:07 PM]: OS: MacOSX 10.12.5, 8 cores, ~ 16 GB RAM. Python: 2.7.12
[02:35:07 PM]: AMPtk v.0.10.0, USEARCH v9.2.64, VSEARCH v2.4.3
[02:35:07 PM]: Loading FASTA Records
[02:35:07 PM]: 1,071 OTUs
[02:35:07 PM]: Appending taxonomy to OTU table and OTUs
[02:35:08 PM]: Generating phylogenetic tree
[02:35:11 PM]: Taxonomy finished: ion.taxonomy.txt
[02:35:11 PM]: Classic OTU table with taxonomy: ion3.otu_table.taxonomy.txt
[02:35:12 PM]: BIOM OTU table created: ion3.biom
[02:35:12 PM]: OTUs with taxonomy: ion3.otus.taxonomy.fa
[02:35:12 PM]: OTU phylogeny: ion3.tree.phy
-------------------------------------------------------
-------------------------------------------------------
[02:35:13 PM]: OS: MacOSX 10.12.5, 8 cores, ~ 16 GB RAM. Python: 2.7.12
[02:35:13 PM]: AMPtk v.0.10.0, USEARCH v9.2.64, VSEARCH v2.4.3
[02:35:13 PM]: Loading FASTA Records
[02:35:13 PM]: 1,115 OTUs
[02:35:13 PM]: Appending taxonomy to OTU table and OTUs
[02

In [58]:
#now convert all to meta transposition for R (vegan) scripts
!amptk meta -i ion3_raw.otu_table.taxonomy.txt -m all_data.meta.csv -o ion3_raw.meta.csv
!amptk meta -i ion3.otu_table.taxonomy.txt -m all_data.meta.csv -o ion3_filtered.meta.csv
!amptk meta -i ion5_raw.otu_table.taxonomy.txt -m all_data.meta.csv -o ion5_raw.meta.csv
!amptk meta -i ion5.otu_table.taxonomy.txt -m all_data.meta.csv -o ion5_filtered.meta.csv
!amptk meta -i miseq4a_raw.otu_table.taxonomy.txt -m all_data.meta.csv -o miseq4a_raw.meta.csv
!amptk meta -i miseq4a.otu_table.taxonomy.txt -m all_data.meta.csv -o miseq4a_filtered.meta.csv
!amptk meta -i miseq4c_raw.otu_table.taxonomy.txt -m all_data.meta.csv -o miseq4c_raw.meta.csv
!amptk meta -i miseq4c.otu_table.taxonomy.txt -m all_data.meta.csv -o miseq4c_filtered.meta.csv
!amptk meta -i miseq6_raw.otu_table.taxonomy.txt -m all_data.meta.csv -o miseq6_raw.meta.csv
!amptk meta -i miseq6.otu_table.taxonomy.txt -m all_data.meta.csv -o miseq6_filtered.meta.csv


Working on complete OTU table: found 1158 OTUs
ERROR m4C-301-1, m4C-301-2, m4C-500-1, m4C-500-2, m4C-712-1, m4C-712-2, m4C-736-1, m4C-736-2, m4C-744-1, m4C-744-2, m4C-755-1, m4C-755-2, m4C-757-1, m4C-757-2, m4C-766-1, m4C-766-2, m4C-mock3-16000, m4C-mock3-32000a, m4C-mock3-32000b, m4C-stds, m4A-301-1, m4A-301-2, m4A-500-1, m4A-500-2, m4A-712-1, m4A-712-2, m4A-736-1, m4A-736-2, m4A-744-1, m4A-744-2, m4A-755-1, m4A-755-2, m4A-757-1, m4A-757-2, m4A-766-1, m4A-766-2, m4A-mock3-16000, m4A-mock3-32000a, m4A-mock3-32000b, m4A-stds, m3-755-2, m3-mock3-3200a, m3-mock3-16000, m3-mock3-32000b, m3-stds, m5-301-1, m5-301-2, m5-500-1, m5-500-2, m5-712-1, m5-712-2, m5-736-1, m5-736-2, m5-744-1, m5-744-2, m5-755-1, m5-755-2, m5-757-1, m5-757-2, m5-766-1, m5-766-2, m5-SynMock1-8000, m5-SynMock2-8000, m5-mock3.0-16000, m5-mock3.5-16000, m6-301-1, m6-301-2, m6-500-1, m6-500-2, m6-712-1, m6-712-2, m6-736-1, m6-736-2, m6-744-1, m6-744-2, m6-755-1, m6-755-2, m6-757-1, m6-757-2, m6-766-1, m6-766-2, m6-mock3-

In [196]:
#map the miseq reads to the mock community fasta file
!vsearch --fastq_filter miseq.fq --fastaout miseq.fa
!vsearch --usearch_global miseq.fa --id 0.97 --db /Users/jon/amptk/DB/amptk_mock2.fa --otutabout miseq.mock.otu_table.txt
!rm miseq.fa

vsearch v2.4.2_osx_x86_64, 16.0GB RAM, 8 cores
https://github.com/torognes/vsearch

Reading input file 100%                                                                                                                                                                                                                                                                                                                                                                                                              
18273964 sequences kept (of which 0 truncated), 0 sequences discarded.
vsearch v2.4.2_osx_x86_64, 16.0GB RAM, 8 cores
https://github.com/torognes/vsearch

Reading file /Users/jon/amptk/DB/amptk_mock2.fa 100%  
14847 nt in 26 seqs, min 335, max 680, avg 571
Masking 100% 
Counting unique k-mers 100% 
Creating index of unique k-mers 100% 
Searching 100%                                                                                                                                              

In [ ]:
#now split OTU table into the different parts/segments needed for different figures
#lets do this in python/pandas
import pandas as pd
from natsort import natsorted

#now load OTU table into dataframe
df = pd.read_csv('miseq.mock.otu_table.txt', sep='\t', index_col=0)
#print df.columns.values
#headers for 
figureOne = ['m3-stds','m3-mock3-16000','m3-mock3-3200a','m3-mock3-32000b',
             'm4A-stds','m4A-mock3-16000','m4A-mock3-32000a','m4A-mock3-32000b',
             'm4C-stds','m4C-mock3-16000','m4C-mock3-32000a','m4C-mock3-32000b',
            'm6-stds', 'm6-mock3-16000', 'm6-mock3-32000a', 'm6-mock3-32000b']

mocks = []
for x in df.index:
    if x.startswith('mock'):
        mocks.append(x)
print mocks
fig1 = df.reindex(columns=figureOne, index=natsorted(mocks))
fig1 = fig1.fillna(value=0).astype(int)
#print fig1
#write to file
fig1.to_csv('synmock.figure1.abundances.txt', sep='\t')
#then generate heatmap
!amptk heatmap -i synmock.figure1.abundances.txt -d tsv -o figure1.pdf -m heatmap --annotate --color YlOrRd --vmax 50000

In [42]:
import pandas as pd
from natsort import natsorted
#load the miseq table, no filtering version
df = pd.read_csv('miseq.sorted.txt', sep='\t', index_col=0)
#Figure 2 should be about index bleed, so generate a table containing the samples that were run in all runs
figureTwo = ['m4A-stds','m4A-mock3-16000','m4A-mock3-32000a','m4A-mock3-32000b','m4A-755-1','m4A-755-2',
             'm4A-744-1','m4A-744-2','m4A-766-1','m4A-766-2',
             'm4b-744-1','m4b-744-2','m4b-766-1','m4b-766-2' ]
df2 = df.reindex(columns=figureTwo, index=natsorted(df.index))
mocks = []
for x in df2.index:
    if x.startswith('mock'):
        mocks.append(x)
otus = [item for item in df2.index if item not in mocks]

df3 = df2.reindex(columns=figureTwo, index=mocks)
df3.to_csv('figure2.onlymocks.txt', sep='\t')
df4 = df2.reindex(columns=figureTwo, index=otus)
#get number of OTUs from df4 as well as total reads
reads = df4.sum(axis=0)
reads.name ="Other OTUs"
otu_count = df4[df4 > 0].count(axis=0, numeric_only=True)
otu_count.name = "Number of OTUs"
#print reads
print otu_count
#now append the read numbers to df3 dataframe
df3 = df3.append(reads)
print df3
total_reads = df3.sum(axis=0)
print total_reads
df3.to_csv('figure2.miseqAB_only2.txt', sep='\t')


!amptk heatmap -i figure2.miseqAB_only2.txt -d tsv -o figure2-2.pdf --format pdf -m heatmap --annotate --color YlOrBr --vmax 30000

m4A-stds             31
m4A-mock3-16000      18
m4A-mock3-32000a     25
m4A-mock3-32000b     33
m4A-755-1           375
m4A-755-2           353
m4A-744-1           281
m4A-744-2           361
m4A-766-1           445
m4A-766-2           276
m4b-744-1           239
m4b-744-2           315
m4b-766-1           349
m4b-766-2           235
Name: Number of OTUs, dtype: int64
                                                    m4A-stds  m4A-mock3-16000  \
#OTU ID                                                                         
mock1 |FP170182|Phialocephala fusca_pident=99.6...      8655              727   
mock2 |RF10JR|unknown Ascomycete_pident=100.0_O...      9176            20769   
mock3 |FP170134|Phialocephala lagerbergii_piden...      8652            28522   
mock4 |FP170235|unknown Ascomycete_pident=100.0...      9051            27727   
mock5 |RF1JR|Aspergillus candidus_pident=100.0_...      8989              147   
mock6 |RF3JR|Bjerkandara adusta_pident=100.0_OTU20     13565  

In [45]:
#now want to take a look at synmock distribution in entire miseq6 run, see if bleed is only happening when there
#is a shared index or not.

import pandas as pd
from natsort import natsorted
#load the miseq table, no filtering version
df = pd.read_csv('miseq_syn.sorted.txt', sep='\t', index_col=0)
#Figure 2 should be about index bleed, so generate a table containing the samples that were run in all runs
keepers = ['m6-301-1', 'm6-301-2', 'm6-500-1', 'm6-500-2', 'm6-712-1', 'm6-712-2', 
           'm6-736-1', 'm6-736-2', 'm6-744-1', 'm6-744-2', 'm6-755-1', 'm6-755-2', 
           'm6-757-1', 'm6-757-2', 'm6-766-1', 'm6-766-2', 
           'm6-mock3-16000', 'm6-mock3-32000a', 'm6-mock3-32000b', 'm6-stds', 'm6-SynMock']
df2 = df.reindex(columns=keepers, index=natsorted(df.index))
mocks = []
for x in df2.index:
    if x.startswith('mock'):
        mocks.append(x)
otus = [item for item in df2.index if item not in mocks]

df3 = df2.reindex(columns=keepers, index=mocks)
df3.to_csv('figure2.onlymocks.txt', sep='\t')
df4 = df2.reindex(columns=keepers, index=otus)
#get number of OTUs from df4 as well as total reads
reads = df4.sum(axis=0)
reads.name ="Other OTUs"
otu_count = df4[df4 > 0].count(axis=0, numeric_only=True)
otu_count.name = "Number of OTUs"
#print reads
print otu_count
#now append the read numbers to df3 dataframe
df3 = df3.append(reads)
#print df3
total_reads = df3.sum(axis=0)
print total_reads
df3.to_csv('figure3.miseq_synmock.txt', sep='\t')


!amptk heatmap -i figure3.miseq_synmock.txt -d tsv -o figure3.pdf --format pdf -m heatmap --annotate --color YlOrBr --vmax 120000

m6-301-1           732
m6-301-2           657
m6-500-1           605
m6-500-2           658
m6-712-1           625
m6-712-2           708
m6-736-1           475
m6-736-2           479
m6-744-1           396
m6-744-2           497
m6-755-1           441
m6-755-2           437
m6-757-1           336
m6-757-2           599
m6-766-1           532
m6-766-2           363
m6-mock3-16000      90
m6-mock3-32000a     89
m6-mock3-32000b    102
m6-stds            100
m6-SynMock          99
Name: Number of OTUs, dtype: int64
m6-301-1            713497
m6-301-2            953709
m6-500-1            723217
m6-500-2            718754
m6-712-1            708178
m6-712-2            693457
m6-736-1            817881
m6-736-2            740933
m6-744-1            612510
m6-744-2            561815
m6-755-1            775334
m6-755-2            891265
m6-757-1            733736
m6-757-2            961336
m6-766-1            694195
m6-766-2            602907
m6-mock3-16000      729561
m6-mock3-32000a     875

### Lets check raw versus filtered for treatment effects

In [24]:
!ls | grep '.biom'

ion3.biom
ion3_raw.biom
ion5.biom
ion5_raw.biom
miseq4a.biom
miseq4a_raw.biom
miseq4c.biom
miseq4c_raw.biom
miseq6.biom
miseq6_raw.biom


In [59]:
!$HOME/amptk/util/phyloseq_nmds.R miseq4a_raw.biom miseq4a_raw.tree.phy miseq4a_raw raupcrick
!$HOME/amptk/util/phyloseq_nmds.R miseq4a.biom miseq4a.tree.phy miseq4a raupcrick
!$HOME/amptk/util/phyloseq_nmds.R miseq4c.biom miseq4c.tree.phy miseq4c raupcrick
!$HOME/amptk/util/phyloseq_nmds.R miseq4c_raw.biom miseq4c_raw.tree.phy miseq4c_raw raupcrick
!$HOME/amptk/util/phyloseq_nmds.R miseq6.biom miseq6.tree.phy miseq6 raupcrick
!$HOME/amptk/util/phyloseq_nmds.R miseq6_raw.biom miseq6_raw.tree.phy miseq6_raw raupcrick
!$HOME/amptk/util/phyloseq_nmds.R ion3.biom ion3.tree.phy ion3 raupcrick
!$HOME/amptk/util/phyloseq_nmds.R ion3_raw.biom ion3_raw.tree.phy ion3_raw raupcrick
!$HOME/amptk/util/phyloseq_nmds.R ion5.biom ion5.tree.phy ion5 raupcrick
!$HOME/amptk/util/phyloseq_nmds.R ion5_raw.biom ion5_raw.tree.phy ion5_raw raupcrick
!$HOME/amptk/util/phyloseq_nmds.R miseq4a_raw.biom miseq4a_raw.tree.phy miseq4a_raw bray
!$HOME/amptk/util/phyloseq_nmds.R miseq4a.biom miseq4a.tree.phy miseq4a bray
!$HOME/amptk/util/phyloseq_nmds.R miseq4c.biom miseq4c.tree.phy miseq4c bray
!$HOME/amptk/util/phyloseq_nmds.R miseq4c_raw.biom miseq4c_raw.tree.phy miseq4c_raw bray
!$HOME/amptk/util/phyloseq_nmds.R miseq6.biom miseq6.tree.phy miseq6 bray
!$HOME/amptk/util/phyloseq_nmds.R miseq6_raw.biom miseq6_raw.tree.phy miseq6_raw bray
!$HOME/amptk/util/phyloseq_nmds.R ion3.biom ion3.tree.phy ion3 bray
!$HOME/amptk/util/phyloseq_nmds.R ion3_raw.biom ion3_raw.tree.phy ion3_raw bray
!$HOME/amptk/util/phyloseq_nmds.R ion5.biom ion5.tree.phy ion5 bray
!$HOME/amptk/util/phyloseq_nmds.R ion5_raw.biom ion5_raw.tree.phy ion5_raw bray

Loading required package: phyloseq
Loading required package: methods
Loading required package: ggplot2
Loading required package: plyr
Loading required package: vegan
Loading required package: permute
Loading required package: lattice
This is vegan 2.4-3
Loading required package: RColorBrewer
[1] ‘1.19.1’
[1] ‘2.2.1’
[1] ‘1.8.4’
[1] ‘2.4.3’
[1] ‘1.1.2’
There were 26 warnings (use warnings() to see them)
[1] "LinkerPrimerSequence" "phinchID"             "Treatment"           
[4] "BarcodeSequence"      "ReversePrimer"        "Terrain"             
[1] "Treatment"
phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 2342 taxa and 16 samples ]
sample_data() Sample Data:       [ 16 samples by 6 sample variables ]
tax_table()   Taxonomy Table:    [ 2342 taxa by 7 taxonomic ranks ]
phy_tree()    Phylogenetic Tree: [ 2342 tips and 2341 internal nodes ]
Run 0 stress 0.09775419 
Run 1 stress 0.1187046 
Run 2 stress 0.09561271 
... New best solution
... Procrustes: rmse 0.113

In [61]:
!$HOME/amptk/util/parse_adonis.py stats summary_stats.csv

| Run         | Distance  | Test       | Terrain | Treatment  |
| ----------- | --------- | ---------- | ------- | ---------- |
| ion3        | bray      | Adonis     | 0.0928  | 0.7441     |
| ion3        | bray      | Betadisper | 0.4653  | 0.6512     |
| ion3        | raupcrick | Adonis     | 0.001   | 0.1928     |
| ion3        | raupcrick | Betadisper | 0.0039  | 0.5488     |
| ion3_raw    | bray      | Adonis     | 0.0998  | 0.7336     |
| ion3_raw    | bray      | Betadisper | 0.4579  | 0.6801     |
| ion3_raw    | raupcrick | Adonis     | 0.0008  | 0.2255     |
| ion3_raw    | raupcrick | Betadisper | 0.0001  | 0.589      |
| ion5        | bray      | Adonis     | 0.0313  | 0.4757     |
| ion5        | bray      | Betadisper | 0.8302  | 0.8232     |
| ion5        | raupcrick | Adonis     | 0.0007  | 0.1047     |
| ion5        | raupcrick | Betadisper | 0.0001  | 0.9046     |
| ion5_raw    | bray      | Adonis     | 0.0296  | 0.4791     |
| ion5_raw    | bray      | Betadisper | 0.8312  | 0.8274     |
| ion5_raw    | raupcrick | Adonis     | 0.0004  | 0.2444     |
| ion5_raw    | raupcrick | Betadisper | 0.0001  | 0.823      |
| miseq4a     | bray      | Adonis     | 0.0186  | 0.4973     |
| miseq4a     | bray      | Betadisper | 0.9397  | 0.9681     |
| miseq4a     | raupcrick | Adonis     | 0.0002  | 0.1406     |
| miseq4a     | raupcrick | Betadisper | 0.0001  | 0.5266     |
| miseq4a_raw | bray      | Adonis     | 0.0147  | 0.4904     |
| miseq4a_raw | bray      | Betadisper | 0.9304  | 0.9868     |
| miseq4a_raw | raupcrick | Adonis     | 0.0004  | 0.0751     |
| miseq4a_raw | raupcrick | Betadisper | 0.0001  | 0.4094     |
| miseq4c     | bray      | Adonis     | 0.0161  | 0.5189     |
| miseq4c     | bray      | Betadisper | 0.8757  | 0.9981     |
| miseq4c     | raupcrick | Adonis     | 0.0001  | 0.1175     |
| miseq4c     | raupcrick | Betadisper | 0.0001  | 0.3067     |
| miseq4c_raw | bray      | Adonis     | 0.014   | 0.494      |
| miseq4c_raw | bray      | Betadisper | 0.8587  | 0.9817     |
| miseq4c_raw | raupcrick | Adonis     | 0.0015  | **0.0099** |
| miseq4c_raw | raupcrick | Betadisper | 0.0001  | **0.0082** |
| miseq6      | bray      | Adonis     | 0.0132  | 0.5091     |
| miseq6      | bray      | Betadisper | 0.8215  | 0.9792     |
| miseq6      | raupcrick | Adonis     | 0.0017  | **0.0124** |
| miseq6      | raupcrick | Betadisper | 0.0001  | 0.9145     |
| miseq6_raw  | bray      | Adonis     | 0.014   | 0.5079     |
| miseq6_raw  | bray      | Betadisper | 0.8237  | 0.9906     |
| miseq6_raw  | raupcrick | Adonis     | 0.0011  | 0.0546     |
| miseq6_raw  | raupcrick | Betadisper | 0.001   | **0.0001** |

The miseq6 data here was filtered with SynMock first (0.1%)- that is where we then "create" a treatment effect with the filtering, while there appears to be no treatment effect in the raw data (although it is close). If you index-bleed filter to a higher value (0.28%) then you do not see the treatment effect. This is of course only in raupcrick distance metric.  This is so dumb.